azkeys/coseSigner.go

```diff
@@ -20,32 +20,83 @@ import (
+// sufficient information to verify the signed product (an identifier for the signing key and the
+// public key.)
+type IdentifiableCoseSigner interface {
+	cose.Signer
+	PublicKey() (*ecdsa.PublicKey, error)
+	KeyIdentifier() string
+}
+
+// IdentifiableCoseSignerFactory is for creating IdentifiableCoseSigners. The reason for a factory
+// here is that we can always create a fresh instance, capturing the latest key information at that
+// point in time.
+type IdentifiableCoseSignerFactory interface {
+	NewIdentifiableCoseSigner(ctx context.Context) (IdentifiableCoseSigner, error)
 }
 
-// NewCoseSignerKeyVault creates a new keyvault configuration that signs with ES384
+// KeyVaultCoseSignerFactory creates instances of our Azure KeyVault implementation of
+// IdentifiableCoseSigner. The keyvault configuration is stored on the object and new instances
+// can be created without caller knowledge of it.
+type KeyVaultCoseSignerFactory struct {
+	keyVaultURL string
+	keyName     string
+}
+
+// NewKeyVaultCoseSignerFactory returns a new instance of the factory, storing the keyvault config
+func NewKeyVaultCoseSignerFactory(keyVaultURL string, keyName string) *KeyVaultCoseSignerFactory {
+	return &KeyVaultCoseSignerFactory{
+		keyVaultURL: keyVaultURL,
+		keyName:     keyName,
+	}
+}
+
+// NewIdentifiableCoseSigner creates a new keyvault configuration that signs with ES384
 // using the latest version of the named key
-func NewCoseSignerKeyVault(keyVaultURL string, keyName string) *CoseSignerKeyVault {
-	kv := CoseSignerKeyVault{
-		keyName: keyName,
+func (f *KeyVaultCoseSignerFactory) NewIdentifiableCoseSigner(ctx context.Context) (IdentifiableCoseSigner, error) {
+	kv := KeyVaultCoseSigner{
+		keyName: f.keyName,
 		alg:     keyvault.ES384, // hardwired for the moment, add caller setting when needed
 		KeyVault: &KeyVault{
-			url: keyVaultURL,
+			url: f.keyVaultURL,
 		},
 	}
-	return &kv
+
+	key, err := kv.GetLatestKey(ctx, kv.keyName)
+	if err != nil {
+		return nil, fmt.Errorf("unable to get latest key for %s: %w", kv.keyName, err)
+	}
+
+	if key.Key.Kid == nil {
+		return nil, fmt.Errorf("key ID is nil")
+	}
+
+	kv.key = key
+
+	return &kv, nil
+}
+
+// KeyVaultCoseSigner is the azure keyvault client for interacting with keyvault keys
+// using a cose.Signer interface
+type KeyVaultCoseSigner struct {
+	keyName string
+	alg     keyvault.JSONWebKeySignatureAlgorithm
+	key     keyvault.KeyBundle
+	*KeyVault
 }
 
 // Algorithm gets the cose algorithm for the key
-func (kv *CoseSignerKeyVault) Algorithm() cose.Algorithm {
+func (kv *KeyVaultCoseSigner) Algorithm() cose.Algorithm {
 	return cose.AlgorithmES384
 }
 
+// KeyIdentifier returns the essential information to identify the key, apart from any platform
+// specific format (i.e. the Azure URL.) It takes the form: <key name>/<key version>.
+func (kv *KeyVaultCoseSigner) KeyIdentifier() string {
+	return fmt.Sprintf("%s/%s", GetKeyName(*kv.key.Key.Kid), GetKeyVersion(*kv.key.Key.Kid))
```

Comment:
suggestion: I wonder if we should have a 'provider' hint in here too ? Ie

`azurekv:<keyname>/<keyversion>`

Otherwise if we have things like

Microsoft Trusted Signing service sigs we will have a legacy of implicitly defaulting to azure kv